oauthproxy.go

```diff
@@ -723,3 +754,37 @@ func (p *OAuthProxy) CheckBasicAuth(req *http.Request) (*providers.SessionState,
 	}
 	return nil, fmt.Errorf("%s not in HtpasswdFile", pair[0])
 }
+
+func (p *OAuthProxy) CheckTLSAuth(req *http.Request) (*providers.SessionState, error) {
+	if req.TLS == nil || len(req.TLS.PeerCertificates) == 0 {
+		return nil, nil
+	}
+	if p.ClientCertVerifyOptions == nil {
+		return nil, nil
+	}
+
+	// Use intermediates, if provided, and verify the certificate
+	opts := *p.ClientCertVerifyOptions
+	if opts.Intermediates == nil && len(req.TLS.PeerCertificates) > 1 {
+		opts.Intermediates = x509.NewCertPool()
+		for _, intermediate := range req.TLS.PeerCertificates[1:] {
+			opts.Intermediates.AddCert(intermediate)
+		}
+	}
+	client := req.TLS.PeerCertificates[0]
+	if _, err := client.Verify(opts); err != nil {
+		return nil, err
+	}
+
+	// email addresses can be verified if the CA is trusted
+	for _, email := range client.EmailAddresses {
+		if p.Validator(email) {
+			return &providers.SessionState{User: email, Email: email}, nil
```

Comment:
looks like this allows clients that do not have the specified common name to access